### PR TITLE
Adds power sinks

### PR DESCRIPTION
**Pull request by MaxIsJoe**
part of #6149

CL: [New] added power sinks

### CODE REVIEW

File: UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				{
+					if(data.InData.Data.ActualVoltage > 0)
+					{
+						currentCharge += data.InData.Data.ActualVoltage * chargeAmplifer;
+					}
+				}
+			}
+			else
+			{
+				Logger.LogError($"Unable to find electrical data for {gameObject.ExpensiveName()} at {gameObject.RegisterTile().WorldPosition}!");
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs]:
```suggestion
```

Don't think this log is necessary?

File: UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			// Get data before despawning
+			var worldPos = gameObject.AssumedWorldPosServer();
+			// Despawn the explosive
+			_ = Despawn.ServerSingle(gameObject);
+			Explosion.StartExplosion(worldPos.RoundToInt(), currentCharge * explosionAmplifer);
+		}
+
+		public void CheckForVoltage()
+		{
+			if (isAnchored == false) return;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs]: does this mean it doesn't actually draw power?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs]: i've given up on trying to do that plus it just complicated the explosion calc so right now this is good as it is

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs]: then, you need to note that if there isn't a Resistant source at the end of the cable no current will flow down and there There won't be voltage

**Comment by Bod9001** [comment]
also isn't the balancing factor of a power Sink that the power goes out?

**Comment by MaxIsJoe** [comment]
> also isn't the balancing factor of a power Sink that the power goes out?

I don't know, I personally have never seen a power sink in SS13 at all so I don't know how they behave.
but from my tests :
when you put the sink next to a generator it generates power so fast that i doubt people will have time to react to the power going out before the power sink tearing it and the connecting electrical grid.
The only moments I think it would make a difference is if you're putting the power sink ontop of a cable thats the only thing that's powering an entire department but I think people will use that as a way to metagame and escape the blast radius.

**Comment by MaxIsJoe** [comment]
Also I added the note and removed the debug.log thing, pls review

**Comment by Bod9001** [comment]
From what I remember the power on the entire station goes out, on the ss13 version and, everyone scrambles to find the power sink, sometimes it's just the engineers Forgetting getting to set up the engine, it even turns off APCs

**Comment by AtnerNT** [comment]
From my understanding it essentially stops all power from flowing past it and from flowing anywhere else that isnt to the sink- 100% of the systems power is focused into the sink, which causes all the power outages once the DPs run flat.

**Comment by MaxIsJoe** [comment]
Here's why I stopped trying adding that functionality
https://youtu.be/68_tzzUiTsU

File: UnityProject/Assets/Scripts/Systems/Electricity/FunctionsAndClasses/ElectricityFunctions.cs [hunk 3]
```diff
@@ -230,10 +230,7 @@ public static void WorkOutActualNumbers(IntrinsicElectronicData ElectricItem)
 
 				foreach (var CurrentItem in AnInterestingDictionary)
 				{
-					if (CurrentItem.Value > 0)
-					{
-						Current += CurrentItem.Value;
-					}
+					Current += CurrentItem.Value;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Systems/Electricity/FunctionsAndClasses/ElectricityFunctions.cs]: did you check stuff like APCs and SMES charging?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/Electricity/FunctionsAndClasses/ElectricityFunctions.cs]: Yes, everything seems to be fine so far.
haven't encountered any issues

**Comment by MaxIsJoe** [comment]
Rebased because i'll be trying to work on this again

If the electrical issue seems to be much more complicated I think I'll have to create another separate branch for fixing why tiles and wires dont get their electrical data updated

**Comment by MaxIsJoe** [comment]
Friends, After a long and tedious process of trial and error and with the thanks of the grand wizard that is Bod..

It is done. Please review.

**Comment by corp-0** [comment]
tests never finished

**Comment by MaxIsJoe** [comment]
Alright this should fix tests now

**Comment by NoooneyDude** [comment]
Tests are still broken.

**Comment by MaxIsJoe** [comment]
> Tests are still broken.

image
no they're not?

**Comment by NoooneyDude** [comment]
Weird. I see that now too. But the last commit has an X beside it: x

/shrug

**Comment by MaxIsJoe** [comment]
> Weird. I see that now too. But the last commit has an X beside it: x
> 
> /shrug

Bod re-ran the tests before

**Comment by NoooneyDude** [comment]
It updates the icon for the last commit when you do that. I've restarted the tests to clear it up.

**Comment by MaxIsJoe** [comment]
> It updates the icon for the last commit when you do that. I've restarted the tests to clear it up.

tests still show as passing so it's working

File: UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			if (interaction.UsedObject.Item().HasTrait(CommonTraits.Instance.Screwdriver))
+			{
+				if (isAnchored && isCharging == false) UnAnchor();
+				else Anchor();
+				SoundManager.PlayNetworkedAtPos(CommonSounds.Instance.screwdriver, gameObject.AssumedWorldPosServer());
+				return;
+			}
+			if (interaction.UsedObject.Item().HasTrait(CommonTraits.Instance.Multitool))
+			{
+				Chat.AddExamineMsg(interaction.Performer, $"This device's battery unit is holding {currentCharge.ToEngineering("V")}");
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs:62]: Odd unit to use for charge.

File: UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+			return true;
+		}
+
+		private void Anchor()
+		{
+			isAnchored = true;
+			pickupable.ServerSetCanPickup(false);
+			objectBehaviour.ServerSetPushable(false);
+			ElectricalManager.Instance.electricalSync.StructureChange = true;
+			Chat.AddLocalMsgToChat($"The {gameObject.ExpensiveName()} makes a clicking sound as it <b>anchors</b> to the ground", gameObject);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs:85]: Would be nice to have it actually make the sound as opposed to saying it makes a sound.

**Review by NoooneyDude — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Weapons/PowerSink.cs:62]: voltage makes more sense because in-code thats what im using.

**Review by Bod9001 — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Systems/Electricity/FunctionsAndClasses/ElectricityFunctions.cs (hunk 3 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Systems/Electricity/FunctionsAndClasses/ElectricityFunctions.cs]: might as well that this since it's not needed

**Review by Bod9001 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

**Pull request merged**